### ISSUE TITLE
tsc crash: TypeError: Cannot read property 'flags' of undefined

### ISSUE THREAD

**Issue by ahagelstein** [Microsoft/TypeScript#14238]
**TypeScript Version:**  2.2.1
**Code**

```ts
export function classExtender<TFunction>(superClass: TFunction, _instanceModifier: (instance: any, args: any[]) => void): TFunction {

    return class decoratorFunc extends superClass {

        constructor(...args: any[]) {

            super(...args);
            _instanceModifier(this, args);
        }
    };
}
```

**Expected behavior:**

compile without crash.

**Actual behavior:**

C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:51296
                throw e;
                ^

TypeError: Cannot read property 'flags' of undefined
    at isValidBaseType (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:23170:24)
    at isConstructorType (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:23051:24)
    at getBaseConstructorTypeOfClass (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:23087:105)
    at getBaseTypeVariableOfClass (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:22888:39)
    at getTypeOfFuncClassEnumModule (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:22900:48)
    at getTypeOfSymbol (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:22959:24)
    at checkClassLikeDeclaration (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:34410:30)
    at checkClassExpression (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:34379:13)
    at checkExpressionWorker (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:32255:28)
    at checkExpression (C:\Users\andree\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:32205:42)

**Comment by mhegazy** [comment]
@sandersn can you take a look.

**Comment by sandersn** [comment]
Looks like `getBaseConstraintOfType` returns `undefined` now, which `isConstructorType` does not handle.

**Comment by sandersn** [comment]
Never mind, `getBaseConstraintOfType` could always return undefined. It's the fact that `isConstructorType` added a call to `getBaseConstraintOfType`, but doesn't check whether the constraint is undefined. All the other calls to `getBaseConstraintOfType` check the return value for undefined.

**Comment by sandersn** [comment]
By the way, this code still does not compile with errors. `TFunction` needs a constraint to indicate that it is a valid superclass. See the explanation in the [Typescript 2.2 announcement for an example](https://blogs.msdn.microsoft.com/typescript/2017/02/22/announcing-typescript-2-2/).

**Comment by sandersn** [comment]
Fix is up at #14250
